File: adapters/typescript/src/profiling.ts
import * as fs from "node:fs";
import * as path from "node:path";

interface ProfilePhase {
  name: string;
  duration_ns: number;
}

class AdapterProfiler {
  private readonly outputPath = process.env.LEXICON_ADAPTER_PROFILE ?? "";
  private readonly phases: ProfilePhase[] = [];
  private readonly counts: Record<string, number> = {};

  measure<T>(name: string, operation: () => T): T {
    if (!this.outputPath) return operation();
    const started = process.hrtime.bigint();
    try {
      return operation();
    } finally {
      this.phases.push({ name, duration_ns: Number(process.hrtime.bigint() - started) });
    }
  }

  set(name: string, value: number): void {
    if (this.outputPath) this.counts[name] = value;
  }

  write(): void {
    if (!this.outputPath) return;
    fs.mkdirSync(path.dirname(this.outputPath), { recursive: true });
    fs.writeFileSync(
      this.outputPath,
      `${JSON.stringify({ version: 1, phases: this.phases, counts: this.counts }, null, 2)}\n`,
      "utf8",
    );
  }
}

export const adapterProfile = new AdapterProfiler();

File: adapters/rust/src/profiling.rs
use anyhow::{Context, Result};
use serde_json::json;
use std::collections::BTreeMap;
use std::env;
use std::fs;
use std::path::PathBuf;
use std::sync::{Mutex, OnceLock};
use std::time::Instant;

#[derive(Default)]
struct AdapterProfile {
    path: Option<PathBuf>,
    phases: Vec<serde_json::Value>,
    counts: BTreeMap<String, i64>,
}

static PROFILE: OnceLock<Mutex<AdapterProfile>> = OnceLock::new();

fn profile() -> &'static Mutex<AdapterProfile> {
    PROFILE.get_or_init(|| {
        Mutex::new(AdapterProfile {
            path: env::var_os("LEXICON_ADAPTER_PROFILE").map(PathBuf::from),
            ..AdapterProfile::default()
        })
    })
}

pub(crate) fn measure<T>(name: &str, operation: impl FnOnce() -> T) -> T {
    let enabled = profile()
        .lock()
        .expect("adapter profile lock")
        .path
        .is_some();
    if !enabled {
        return operation();
    }
    let started = Instant::now();
    let result = operation();
    profile()
        .lock()
        .expect("adapter profile lock")
        .phases
        .push(json!({
            "name": name,
            "duration_ns": started.elapsed().as_nanos() as u64,
        }));
    result
}

pub(crate) fn set(name: &str, value: usize) {
    let mut profile = profile().lock().expect("adapter profile lock");
    if profile.path.is_some() {
        profile.counts.insert(name.to_string(), value as i64);
    }
}

pub(crate) fn write() -> Result<()> {
    let profile = profile().lock().expect("adapter profile lock");
    let Some(path) = profile.path.as_ref() else {
        return Ok(());
    };
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)
            .with_context(|| format!("create adapter profile directory {}", parent.display()))?;
    }
    let mut data = serde_json::to_vec_pretty(&json!({
        "version": 1,
        "phases": profile.phases,
        "counts": profile.counts,
    }))?;
    data.push(b'\n');
    fs::write(path, data).with_context(|| format!("write adapter profile {}", path.display()))?;
    Ok(())
}
